docs/contribute/testing.adoc

```diff
@@ -0,0 +1,423 @@
+
+
+== Preparing test environment and running tests
+To run a specific test or all tests you first need to compile the OpenSCAP
+library and then install additional packages required for testing. See the
+*Compilation* section in the link:../../README.md[README.md] for more details.
+
+
+== Writing a new test
+In this guide we will use an example to describe the process of writing a test
+for a pull request. Let's suppose that you already have a commit which adds
+a new feature. Let's suppose that this new feature allows to evaluate a single
+rule from a specific profile instead of all the rules in that profile.
+
+NOTE: The single rule evaluation feature was added into the OpenSCAP through
+the commit
+link:https://github.com/OpenSCAP/openscap/commit/4637b132e710d7ab73601718a531a8587b406cd3[4637b13].
+
+To sum up, we are adding the highlighted option to the `xccdf` module of
+the OpenSCAP command line tool:
+[[app-listing]]
+[subs=+quotes]
+----
+$ oscap xccdf eval *--rule X* --profile Y \
+    /usr/share/xml/scap/ssg/content/ssg-fedora-ds.xml
+----
+
+
+=== Deciding where to put a new test
+In our example, we are changing the `xccdf` module, therefore we will look for
+its subdirectory in the link:../../tests[tests] direcotory and we will find it
+at the following link: link:../../tests/API/XCCDF[tests/API/XCCDF]. There are
+multiple other subdirectories with tests and we need to decide where to put our
+test for the evaluation of a single rule. From all these subdirectories the
+most suitable one appears to be the
+link:../../tests/API/XCCDF/unittests[tests/API/XCCDF/unittests] directory as
+all the general XCCDF tests for `oscap xccdf eval` are located there. We will
+add our new test into this subdirectory.
+
+
+==== Scenario A: There is a suitable directory for my new test
+This will happen most of the times. As stated above in our example we will place
+our test into the
+link:../../tests/API/XCCDF/unittests[tests/API/XCCDF/unittests] directory.
+
+
+==== Scenario B: There is no suitable directory for my new test
+This might happen if your test covers a part of the OpenSCAP which has no tests
+at all. In this case you would need to add a new directory with suitable name
+into the link:../../tests[tests] directory structure and create/update
+Makefiles.
+
+To have an example also for this scenario, let's suppose we want to add the
+`foo` subdirectory into the link:../../tests/API/XCCDF[tests/API/XCCDF]
+directory. We need to:
+
+. Create the `foo` subdirectory in the
+   link:../../tests/API/XCCDF[tests/API/XCCDF] directory.
+. Add a new entry for `foo` subdirectory into the `SUBDIRS` variable in
+   the link:../../tests/API/XCCDF/Makefile.am[tests/API/XCCDF/Makefile.am].
+   When adding a new entry please keep in mind that all entries are sorted in
+   the alphabetical order.
+. Create `Makefile.am` inside the `tests/API/XCCDF/foo` directory with the
+   following contents:
+[[app-listing]]
+[source,bash]
```

Comment:
is explaining autotools really in scope? there are better resources for this, perhaps link autotools docs?